File: JALib/Stream/ByteArrayDataOutput.cs
using System;
using System.Text;
using System.Threading.Tasks;
using JALib.Core;
using JALib.Tools;

namespace JALib.Stream;

public class ByteArrayDataOutput : IDisposable {
    private byte[] buf;
    private int count;
    private JAMod mod;
    
    public ByteArrayDataOutput(JAMod mod = null) {
        buf = new byte[32];
        this.mod = mod;
    }

    public void Dispose() {
        buf = null;
        mod = null;
    }

    private void EnsureCapacity(int minCapacity) {
        int oldCapacity = buf.Length;
        int minGrowth = minCapacity - oldCapacity;
        if(minGrowth <= 0) return;
        byte[] bytes = new byte[Math.Max(buf.Length + 16, minCapacity)];
        Array.Copy(buf, bytes, count);
        buf = bytes;
    }
    
    public void WriteUTF(string value) {
        WriteBytes(Encoding.UTF8.GetBytes(value));
    }

    public void WriteInt(int value) {
        EnsureCapacity(count + 4);
        WriteIntBypass(value);
    }

    private void WriteIntBypass(int value) {
        buf[count++] = (byte) (value >> 24);
        buf[count++] = (byte) (value >> 16);
        buf[count++] = (byte) (value >> 8);
        buf[count++] = (byte) value;
    }

    public void WriteULong(ulong value) {
        EnsureCapacity(count + 8);
        buf[count++] = (byte) (value >> 56);
        buf[count++] = (byte) (value >> 48);
        buf[count++] = (byte) (value >> 40);
        buf[count++] = (byte) (value >> 32);
        buf[count++] = (byte) (value >> 24);
        buf[count++] = (byte) (value >> 16);
        buf[count++] = (byte) (value >> 8);
        buf[count++] = (byte) value;
    }

    public void WriteLong(long value) {
        EnsureCapacity(count + 8);
        buf[count++] = (byte) (value >> 56);
        buf[count++] = (byte) (value >> 48);
        buf[count++] = (byte) (value >> 40);
        buf[count++] = (byte) (value >> 32);
        buf[count++] = (byte) (value >> 24);
        buf[count++] = (byte) (value >> 16);
        buf[count++] = (byte) (value >> 8);
        buf[count++] = (byte) value;
    }

    public void WriteBoolean(bool value) {
        EnsureCapacity(count + 1);
        buf[count++] = value ? (byte) 1 : (byte) 0;
    }

    public void WriteFloat(float value) {
        EnsureCapacity(count + 4);
        byte[] buffer = BitConverter.GetBytes(value);
        buf[count++] = buffer[3];
        buf[count++] = buffer[2];
        buf[count++] = buffer[1];
        buf[count++] = buffer[0];
    }

    public void WriteDouble(double value) {
        EnsureCapacity(count + 8);
        byte[] buffer = BitConverter.GetBytes(value);
        buf[count++] = buffer[7];
        buf[count++] = buffer[6];
        buf[count++] = buffer[5];
        buf[count++] = buffer[4];
        buf[count++] = buffer[3];
        buf[count++] = buffer[2];
        buf[count++] = buffer[1];
        buf[count++] = buffer[0];
    }

    public void WriteByte(byte value) {
        EnsureCapacity(count + 1);
        buf[count++] = value;
    }

    public void WriteShort(short value) {
        EnsureCapacity(count + 2);
        buf[count++] = (byte) (value >> 8);
        buf[count++] = (byte) value;
    }

    public void WriteBytes(byte[] value) {
        EnsureCapacity(count + value.Length + 4);
        WriteIntBypass(value.Length);
        foreach(byte b in value) buf[count++] = b;
    }
    
    public void WriteUShort(ushort value) {
        EnsureCapacity(count + 2);
        buf[count++] = (byte) (value >> 8);
        buf[count++] = (byte) value;
    }
    
    public void WriteUInt(uint value) {
        EnsureCapacity(count + 4);
        buf[count++] = (byte) (value >> 24);
        buf[count++] = (byte) (value >> 16);
        buf[count++] = (byte) (value >> 8);
        buf[count++] = (byte) value;
    }
    
    public void WriteSByte(sbyte value) {
        EnsureCapacity(count + 1);
        buf[count++] = (byte) value;
    }
    
    public void WriteChar(char value) {
        EnsureCapacity(count + 2);
        buf[count++] = (byte) (value >> 8);
        buf[count++] = (byte) value;
    }
    
    public void WriteDecimal(decimal value) {
        int[] bits = decimal.GetBits(value);
        EnsureCapacity(count + bits.Length * 4);
        foreach(int i in bits) WriteIntBypass(i);
    }

    public byte[] ToByteArray() {
        byte[] data = new byte[count];
        Array.Copy(buf, data, count);
        Dispose();
        return data;
    }
}

File: JALib-Server/src/main/java/kr/jongyeol/jaServer/packet/ByteArrayDataOutput.java
package kr.jongyeol.jaServer.packet;

import lombok.SneakyThrows;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;

public class ByteArrayDataOutput {
    private byte[] buf;
    private int count = 0;

    public ByteArrayDataOutput() {
        buf = new byte[32];
    }

    private void ensureCapacity(int minCapacity) {
        int oldCapacity = buf.length;
        int minGrowth = minCapacity - oldCapacity;
        if(minGrowth > 0) buf = Arrays.copyOf(buf, Math.max(buf.length + 16, minCapacity));
    }

    public void writeUTF(String value) {
        writeBytes(value == null ? null : value.getBytes(StandardCharsets.UTF_8));
    }

    public void writeInt(int value) {
        ensureCapacity(count + 4);
        writeIntBypass(value);
    }

    private void writeIntBypass(int value) {
        buf[count++] = (byte) (value >>> 24);
        buf[count++] = (byte) (value >>> 16);
        buf[count++] = (byte) (value >>> 8);
        buf[count++] = (byte) (value);
    }

    public void writeLong(long value) {
        ensureCapacity(count + 8);
        buf[count++] = (byte) (value >>> 56);
        buf[count++] = (byte) (value >>> 48);
        buf[count++] = (byte) (value >>> 40);
        buf[count++] = (byte) (value >>> 32);
        buf[count++] = (byte) (value >>> 24);
        buf[count++] = (byte) (value >>> 16);
        buf[count++] = (byte) (value >>> 8);
        buf[count++] = (byte) (value);
    }

    public void writeBoolean(boolean value) {
        ensureCapacity(count + 1);
        buf[count++] = value ? (byte) 1 : 0;
    }

    public void writeFloat(float value) {
        writeInt(Float.floatToIntBits(value));
    }

    public void writeDouble(double value) {
        writeLong(Double.doubleToLongBits(value));
    }

    @SneakyThrows
    public void writeByte(byte value) {
        ensureCapacity(count + 1);
        buf[count++] = value;
    }

    public void writeShort(short value) {
        ensureCapacity(count + 2);
        buf[count++] = (byte) (value >>> 8);
        buf[count++] = (byte) (value);
    }

    @SneakyThrows
    public void writeBytes(byte[] value) {
        if(value == null) {
            writeInt(-1);
            return;
        }
        ensureCapacity(count + value.length + 4);
        writeIntBypass(value.length);
        for(byte v : value) buf[count++] = v;
    }

    public void writeBytesOnly(byte[] value) {
        ensureCapacity(count + value.length);
        for(byte v : value) buf[count++] = v;
    }

    public void writeBytes(byte[] value, int offset, int length) {
        ensureCapacity(count + length + 4);
        writeIntBypass(length);
        for(int i = offset; i < offset + length; i++) buf[count++] = value[i];
    }

    public byte[] toByteArray() {
        return Arrays.copyOf(buf, count);
    }

    public void close() {
        buf = null;
    }
}
